File: frontend/src/api/retiradaLojas.ts
import axios from "axios";

const isProduction = (): boolean => {
  if (typeof window === "undefined") return false;
  const hostname = window.location.hostname;
  return (
    hostname.includes("vercel.app") ||
    hostname.includes("netlify.app") ||
    import.meta.env.PROD ||
    (hostname !== "localhost" && hostname !== "127.0.0.1")
  );
};

const normalizeApiUrl = (url: string | undefined): string => {
  const envUrl = url || import.meta.env.VITE_API_BASE_URL;
  if (envUrl) {
    let normalizedUrl = envUrl.trim().replace(/\/+$/, "");
    if (!normalizedUrl.match(/^https?:\/\//i)) {
      normalizedUrl = `https://${normalizedUrl}`;
    }
    return normalizedUrl;
  }
  if (isProduction()) {
    return "https://automacao-relatorio-production.up.railway.app";
  }
  return "http://127.0.0.1:8000";
};

const API_BASE_URL = normalizeApiUrl(import.meta.env.VITE_API_BASE_URL);

// ── Tipos base ────────────────────────────────────────────────────────────────

export interface ProductLookupResult {
  codigo_linx: string;
  descricao: string | null;
  preco_custo: number | null;
  codigo_barras: string | null;
  sku: string | null;
}

export interface RetiradaItemCreate {
  codigo_linx: string | null;
  codigo_barras: string | null;
  descricao: string | null;
  preco_custo: number | null;
}

export interface RetiradaCreate {
  loja: string;
  responsavel: string;
  observacao?: string;
  items: RetiradaItemCreate[];
}

export interface RetiradaItemResponse {
  id: number;
  retirada_id: number;
  codigo_linx: string | null;
  codigo_barras: string | null;
  descricao: string | null;
  preco_custo: number | null;
}

// ── Status ────────────────────────────────────────────────────────────────────

export type RetiradaStatus = "pendente" | "aprovado" | "rejeitado";

// ── Resposta completa (com itens) ─────────────────────────────────────────────

export interface RetiradaResponse {
  id: number;
  loja: string;
  responsavel: string;
  observacao: string | null;
  created_at: string;
  items: RetiradaItemResponse[];
  // Aprovação
  status: RetiradaStatus;
  aprovado_por: string | null;
  aprovado_em: string | null;
  motivo_rejeicao: string | null;
  // Robô Playwright (integração futura)
  robo_executado: boolean;
  robo_executado_em: string | null;
  robo_resultado: string | null;
}

// ── Item de listagem (sem itens detalhados) ───────────────────────────────────

export interface RetiradaListItem {
  id: number;
  loja: string;
  responsavel: string;
  created_at: string;
  total_items: number;
  // Aprovação
  status: RetiradaStatus;
  aprovado_por: string | null;
  aprovado_em: string | null;
  motivo_rejeicao: string | null;
  // Robô Playwright (integração futura)
  robo_executado: boolean;
  robo_executado_em: string | null;
  robo_resultado: string | null;
}

// ── Funções de API ────────────────────────────────────────────────────────────

/** Busca produto pelo código de barras (bipagem) */
export const lookupProductByBarcode = async (
  barcode: string
): Promise<ProductLookupResult> => {
  const response = await axios.get<ProductLookupResult>(
    `${API_BASE_URL}/api/retirada-lojas/lookup/barcode/${encodeURIComponent(barcode)}`
  );
  return response.data;
};

/** Cria uma nova retirada com os itens bipados */
export const createRetirada = async (
  payload: RetiradaCreate
): Promise<RetiradaResponse> => {
  const response = await axios.post<RetiradaResponse>(
    `${API_BASE_URL}/api/retirada-lojas/`,
    payload
  );
  return response.data;
};

/** Lista todas as retiradas (com filtros opcionais) */
export const listRetiradas = async (params?: {
  loja?: string;
  status_filter?: RetiradaStatus;
}): Promise<RetiradaListItem[]> => {
  const response = await axios.get<RetiradaListItem[]>(
    `${API_BASE_URL}/api/retirada-lojas/`,
    { params }
  );
  return response.data;
};

/** Busca uma retirada completa por ID (inclui todos os itens) */
export const getRetirada = async (id: number): Promise<RetiradaResponse> => {
  const response = await axios.get<RetiradaResponse>(
    `${API_BASE_URL}/api/retirada-lojas/${id}`
  );
  return response.data;
};

/** Aprova uma retirada pendente */
export const aprovarRetirada = async (
  id: number,
  aprovadoPor: string
): Promise<RetiradaResponse> => {
  const response = await axios.patch<RetiradaResponse>(
    `${API_BASE_URL}/api/retirada-lojas/${id}/aprovar`,
    { aprovado_por: aprovadoPor }
  );
  return response.data;
};

/** Rejeita uma retirada pendente com motivo */
export const rejeitarRetirada = async (
  id: number,
  motivo: string
): Promise<RetiradaResponse> => {
  const response = await axios.patch<RetiradaResponse>(
    `${API_BASE_URL}/api/retirada-lojas/${id}/rejeitar`,
    { motivo_rejeicao: motivo }
  );
  return response.data;
};

File: backend/app/api/retirada_lojas.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from app.database import get_db
from app.models.retirada_loja import RetiradaLoja, RetiradaLojaItem
from app.schemas.retirada_loja import (
    RetiradaLojaCreate,
    RetiradaLojaResponse,
    RetiradaLojaListItem,
    ProductLookupResponse,
)
from app.services.product_service import get_product_by_codigo_barras

router = APIRouter(prefix="/api/retirada-lojas", tags=["retirada_lojas"])


@router.get("/lookup/barcode/{barcode}", response_model=ProductLookupResponse)
async def lookup_product_by_barcode(barcode: str, db: Session = Depends(get_db)):
    """Busca produto pelo código de barras para uso na bipagem"""
    product = get_product_by_codigo_barras(db, barcode)
    if not product:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Produto não encontrado para o código de barras: {barcode}",
        )
    return ProductLookupResponse(
        codigo_linx=product.codigo_linx,
        descricao=product.descricao,
        preco_custo=float(product.preco_custo) if product.preco_custo is not None else None,
        codigo_barras=product.codigo_barras,
        sku=product.sku,
    )


@router.get("/", response_model=List[RetiradaLojaListItem])
async def list_retiradas(
    skip: int = 0,
    limit: int = 100,
    loja: Optional[str] = None,
    db: Session = Depends(get_db),
):
    """Lista retiradas para lojas salvas"""
    query = db.query(RetiradaLoja)
    if loja:
        query = query.filter(RetiradaLoja.loja == loja)
    retiradas = query.order_by(RetiradaLoja.created_at.desc()).offset(skip).limit(limit).all()
    return [
        RetiradaLojaListItem(
            id=r.id,
            loja=r.loja,
            responsavel=r.responsavel,
            created_at=r.created_at,
            total_items=len(r.items),
        )
        for r in retiradas
    ]


@router.get("/{retirada_id}", response_model=RetiradaLojaResponse)
async def get_retirada(retirada_id: int, db: Session = Depends(get_db)):
    """Busca uma retirada por ID com todos os seus itens"""
    retirada = db.query(RetiradaLoja).filter(RetiradaLoja.id == retirada_id).first()
    if not retirada:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Retirada {retirada_id} não encontrada",
        )
    return retirada


@router.post("/", response_model=RetiradaLojaResponse, status_code=status.HTTP_201_CREATED)
async def create_retirada(payload: RetiradaLojaCreate, db: Session = Depends(get_db)):
    """Salva uma retirada para loja com todos os itens bipados"""
    if not payload.items:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="A retirada deve conter pelo menos um item.",
        )

    retirada = RetiradaLoja(
        loja=payload.loja,
        responsavel=payload.responsavel,
        observacao=payload.observacao,
    )
    db.add(retirada)
    db.flush()

    for item_data in payload.items:
        item = RetiradaLojaItem(
            retirada_id=retirada.id,
            codigo_linx=item_data.codigo_linx,
            codigo_barras=item_data.codigo_barras,
            descricao=item_data.descricao,
            preco_custo=item_data.preco_custo,
        )
        db.add(item)

    db.commit()
    db.refresh(retirada)
    return retirada
